Makefile

```diff
@@ -63,9 +63,65 @@ vet: ## Run go vet against code.
 	go vet ./...
 
 .PHONY: test
-test: manifests generate fmt vet envtest ## Run tests.
+test: manifests generate fmt vet $(ENVTEST) ## Run tests.
 	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test ./... -coverprofile cover.out
 
+.PHONY: require-%
+require-%:
+	@if ! command -v $* 1> /dev/null 2>&1; then echo "$* not found in \$$PATH"; exit 1; fi
+
+ARTIFACT_DIR ?= .test
+
+.PHONY: test-e2e
+test-e2e: $(ARTIFACT_DIR)/kind.kubeconfig kcp-synctarget ready-deployment
+	go test ./test/e2e/...
+
+.PHONY: ready-deployment
+ready-deployment: KUBECONFIG = $(ARTIFACT_DIR)/kcp.kubeconfig
+ready-deployment: kind-image install deploy
+	$(KCP_KUBECTL) --namespace "controller-runtime-example-system" rollout status deployment/controller-runtime-example-controller-manager
+
+.PHONY: kind-image
+kind-image: docker-build
+	kind load docker-image $(IMG) --name controller-runtime-example
+
+$(ARTIFACT_DIR)/kind.kubeconfig: $(ARTIFACT_DIR) require-kind
```

Comment:
@sttts how to debug why this runs always even when the output file exists would be sweet